### DISCUSSION TITLE
Why websocket depends on future crate?

### DISCUSSION THREAD

**Discussion by doowonee** [tokio-rs/axum#1257] [Q&A] [answered]
Hello maybe It is about rust asnyc echo system but I have a question.

Tokio does not support `Sink(future crate)` trait from here <https://github.com/tokio-rs/tokio/discussions/4909>. But Stream(future crate) can created by `tokio-stream` crate.

When I split websocket and it gives me Sink and Stream. But tokio does not support `forward()` function. So I have to import `future crate`

I wonder why axum uses Sink and Stream which depends on `future crate`. Is it impossible to remove dependency future crate and only use tokio crate?

**Comment by davidpdrsn** [comment, marked as answer]
That's impossible to remove. That's just what async rust is today.

**Reply by doowonee** [reply]
So future crate is essential in rust async echo system even more than tokio which provide async runtime. I understand Thank you!

**Reply by davidpdrsn** [reply]
Tokio provides the runtime. Futures provides many utilities and abstractions.
